File: src/main/java/sh/ball/parser/svg/MoveTo.java
package sh.ball.parser.svg;

import java.util.ArrayList;
import java.util.List;

import sh.ball.shapes.Shape;
import sh.ball.shapes.Vector2;

class MoveTo {

  // Parses moveto commands (M and m commands)
  private static List<Shape> parseMoveTo(SvgState state, List<Float> args, boolean isAbsolute) {
    if (args.size() % 2 != 0 || args.size() < 2) {
      for (Float arg : args) {
        System.out.println(arg);
      }
      throw new IllegalArgumentException("SVG moveto command has incorrect number of arguments.\n");
    }

    Vector2 vec = new Vector2(args.get(0), args.get(1));

    if (isAbsolute) {
      state.currPoint = vec;
      state.initialPoint = state.currPoint;
      if (args.size() > 2) {
        return LineTo.absolute(state, args.subList(2, args.size()));
      }
    } else {
      state.currPoint = state.currPoint.translate(vec);
      state.initialPoint = state.currPoint;
      if (args.size() > 2) {
        return LineTo.relative(state, args.subList(2, args.size()));
      }
    }

    return new ArrayList<>();
  }

  static List<Shape> absolute(SvgState state, List<Float> args) {
    return parseMoveTo(state, args, true);
  }

  static List<Shape> relative(SvgState state, List<Float> args) {
    return parseMoveTo(state, args, false);
  }
}

File: Source/svg/MoveTo.cpp
#include "MoveTo.h"
#include "LineTo.h"

std::vector<std::unique_ptr<Shape>> MoveTo::absolute(SvgState& state, std::vector<float>& args) {
	return parseMoveTo(state, args, true);
}

std::vector<std::unique_ptr<Shape>> MoveTo::relative(SvgState& state, std::vector<float>& args) {
	return parseMoveTo(state, args, false);
}

std::vector<std::unique_ptr<Shape>> MoveTo::parseMoveTo(SvgState& state, std::vector<float>& args, bool isAbsolute) {
    if (args.size() % 2 != 0 || args.size() < 2) {
        return {};
    }

    Vector2 vec(args[0], args[1]);

    if (isAbsolute) {
        state.currPoint = vec;
        state.initialPoint = state.currPoint;
        if (args.size() > 2) {
			std::vector<float> newArgs = std::vector<float>(args.begin() + 2, args.end());
            return LineTo::absolute(state, newArgs);
        }
    } else {
		state.currPoint.translate(vec.x, vec.y);
        state.initialPoint = state.currPoint;
        if (args.size() > 2) {
            std::vector<float> newArgs = std::vector<float>(args.begin() + 2, args.end());
            return LineTo::relative(state, newArgs);
        }
    }

    return {};
}
